### ISSUE TITLE
Add e2e tests for .NET

### ISSUE THREAD

**Issue by eyalb181** [metalbear-co/mirrord#439]
We should add e2e tests for common .NET workflows.

**Comment by IsaacCloos** [comment]
# ✋🏻

**Comment by eyalb181** [comment]
Good luck!

**Comment by IsaacCloos** [comment]
# Preamble

Sorry for delay! I have most of my free time on the weekends, and was only just able to sit down to get a lay-of-the-land on this issue. If at any point this issue or related works need to be escalated please reassign 👍🏻. 

[TLDR](#question)

# Stage 

| Stage | Status | 
| :---         |     :---:      |   
| Get Current Tests Working   |  In Progress  | 
| Define Test Expectations     |       | 
| Replicate Tests in .NET | |

# Issues
System details
> macOS Monterey Version 12.5 | Arch : arm64

## test>README.md
https://github.com/metalbear-co/mirrord/blob/3795e759d42713cd97624f1ced62d7f0a31e8817/tests/README.md?plain=1#L1-L11

Following these instructions was more difficult (but doable) than the excellent [online resource](https://mirrord.dev/docs/developer/testing/) you have hosted. I think it would be an easy win to point to that resource from this file. 

I had mixed results getting the [minikube](https://minikube.sigs.k8s.io/docs/) deployment to execute the tests consistently, but after following the online guide it was smooth sailing (minus the points to follow). A possibly helpful upgrade to the doc site would be extending the examples to include minikube (since it's specifically recommended in the test readme file above).

## Flakey tests

#473 will be helpful on my end absolutely. I've had mixed results from this test regardless of configuration or local kube solution. Typically related to www.live.com (which forwards to [outlook](https://outlook.live.com/owa/)), but often others as well.

https://github.com/metalbear-co/mirrord/blob/3795e759d42713cd97624f1ced62d7f0a31e8817/tests/src/sanity.rs#L1049-L1062

## Failing tests

All Golang related e2e tests fail on my local machine with similar errors to this:

```
Launching "go-e2e-dns/19" with arguments []

Extracted library file to "/var/folders/xx/5c4wz4613s780646p68l6g380000gn/T/test.ut6bK01O1C3G/x0slmsfwds-libmirrord_layer.dylib"

Couldn't execute Errno(Errno { code: 2, description: Some("No such file or directory") })
```

I have spent this morning trying to figure out if these errors are a fault of how I followed the test environment instructions with no luck. I have found related work #248 #260, but I'm not entirely sure these are justifications for me to ignore these failing on my machine, so I thought I'd ask. One other piece of evidence I can point to is:

https://github.com/metalbear-co/mirrord/blob/3795e759d42713cd97624f1ced62d7f0a31e8817/.github/workflows/ci.yaml#L152-L279

All of the go-e2e tests are omitted from the macOS ci test suite.

# Question

Having followed the [developer testing guide](https://mirrord.dev/docs/developer/testing/):
Should I be concerned that my local environment is not set up correctly if any of these tests fail? I want to make sure these things are all in order before I add new tests and believe them to be consistent, passable, and properly fail-able when real problems arise.

Thanks so much for reading all my blathering! Any feedback/advice would mean the world to me 😄

**Comment by infiniteregrets** [comment]
> Launching "go-e2e-dns/19" with arguments []
Extracted library file to "/var/folders/xx/5c4wz4613s780646p68l6g380000gn/T/test.ut6bK01O1C3G/x0slmsfwds-libmirrord_layer.dylib"
Couldn't execute Errno(Errno { code: 2, description: Some("No such file or directory") })

How are you running mirrord? This happens when the binary path is not valid

**Comment by aviramha** [comment]
> > Launching "go-e2e-dns/19" with arguments []
> > Extracted library file to "/var/folders/xx/5c4wz4613s780646p68l6g380000gn/T/test.ut6bK01O1C3G/x0slmsfwds-libmirrord_layer.dylib"
> > Couldn't execute Errno(Errno { code: 2, description: Some("No such file or directory") })
> 
> How are you running mirrord? This happens when the binary path is not valid

I assume they didn't compile the go e2e tests, we should add a note that this is required for go e2e to run.

**Comment by IsaacCloos** [comment]
> > Launching "go-e2e-dns/19" with arguments []
> > Extracted library file to "/var/folders/xx/5c4wz4613s780646p68l6g380000gn/T/test.ut6bK01O1C3G/x0slmsfwds-libmirrord_layer.dylib"
> > Couldn't execute Errno(Errno { code: 2, description: Some("No such file or directory") })
> 
> How are you running mirrord? This happens when the binary path is not valid

The online [Testing & Development Guide](https://mirrord.dev/docs/developer/testing/#build-and-run-mirrord) recommends running mirrord with a local process:
`MIRRORD_AGENT_IMAGE=test MIRRORD_AGENT_RUST_LOG=debug RUST_LOG=debug target/debug/mirrord exec -c --pod-name py-serv-deployment-ff89b5974-x9tjx node sample/node/app.mjs`. This is what I've been doing (with the pod name changed).

---

> I assume they didn't compile the go e2e tests, we should add a note that this is required for go e2e to run.

I have attempted the tests again after compiling all of the Go folders included in the project with no success. Could you elaborate on how exactly this should be done? 

Thank you both for your feedback and advice 😄

**Comment by aviramha** [comment]
@IsaacCloos - Thank you and sorry for the troubles!
I think it's best to follow what we do in the [CI yaml](https://github.com/metalbear-co/mirrord/blob/main/.github/workflows/ci.yaml) specifically the e2e part.
For each go directory, we both compile it for go 1.18 and go 1.19.

```
      - uses: actions/setup-go@v3
        with:
          go-version: "1.18.0"
      - run: |
          cd tests/go-e2e
          go build -o 18
      - uses: actions/setup-go@v3
        with:
          go-version: "1.19.0"
      - run: |
          cd tests/go-e2e
          go build -o 19
```

**Comment by IsaacCloos** [comment]
Thank you for the help!

I've got a draft PR open for feedback and I'd love to iterate on that work until it meets the needs of this project.

Also, 

The test suite of this project has been a learning curve. The different ways to run the project, container environments to use, and different architectures can lead to a lot of extra work to feel comfortable with tests (new and old) working correctly and reporting relevant failures/successes. With that said, after this issue is closed I'd love to provide feedback to perhaps improve the QOL for future contributors looking to work in the tests section of the project.

**Comment by IsaacCloos** [comment]
Absolutely no rush on this; can the title and description be updated to reflect the newly desired integration tests?

Do these tests live within the `mirrord-layer>tests` exclusively?

I want to make sure these tests are helpful and designed correctly. A little direction would make a huge difference 😄

**Comment by aviramha** [comment]
> Absolutely no rush on this; can the title and description be updated to reflect the newly desired integration tests?
> 
> Do these tests live within the `mirrord-layer>tests` exclusively?
> 
> I want to make sure these tests are helpful and designed correctly. A little direction would make a huge difference 😄

Yes, that's the part in mirrord that is affected by frameworks/languages so it's actually testing it's flow and logic, hence being there. 
You can see example of the other IT tests, but in general the idea is that we run an app with layer injected to it (using LD/DYLD method), connect it to a "mock" agent that validates the requests and sends the appropriate responses to see that it behaves correctly in a deterministic manner.
